### ISSUE TITLE
How to properly incorporate sklearn metrics?

### ISSUE THREAD

**Issue by afogarty85** [horovod/horovod#2509]
I am looking for help as to how to collect sklearn metrics, such as f1, when using Horovod.

Consider the following evaluation fn,
```
def evaluate(model, val_dataloader_iter, validation_steps, device, 
             metric_agg_fn=None):
  model.eval()  # Set model to evaluate mode

  # statistics
  running_loss = 0.0
  running_corrects = 0
  running_f1 = 0.0
  
  # Iterate over all the validation data.
  for step in range(validation_steps):
    x_cat, x_cont, labels = _transform_row(next(val_dataloader_iter))

    # Do not track history in evaluation to save memory
    with torch.set_grad_enabled(False):
      # forward
      logits, probas = model(x_cat.long(), x_cont.squeeze(1))
      _, preds = torch.max(probas, 1)
      loss = F.cross_entropy(logits, labels.long(), weight=weighting)

    # statistics
    running_loss += loss.item()
    running_corrects += torch.sum(preds == labels)
    running_f1 += f1_score(preds, labels, average='weighted')
    
  # Average the losses across observations for each minibatch.
  epoch_loss = running_loss / validation_steps
  epoch_acc = running_corrects.double() / (validation_steps * BATCH_SIZE)
  epoch_f1 = running_f1 / (validation_steps * BATCH_SIZE) # testing
  
  # metric_agg_fn is used in the distributed training to aggregate the metrics on all workers
  if metric_agg_fn is not None:
    epoch_loss = metric_agg_fn(epoch_loss, 'avg_loss')
    epoch_acc = metric_agg_fn(epoch_acc, 'avg_acc')
    epoch_f1 = metric_agg_fn(epoch_f1, 'avg_f1')

  print('Validation Loss: {:.4f} Acc: {:.4f} F1: {:.4f}'.format(epoch_loss, epoch_acc, epoch_f1))
  return epoch_loss, epoch_acc, epoch_f1
```

```
# this is set to call correctly in the Horovod distributed training code, but omitted for clarity here
def metric_average(val, name):
  tensor = torch.tensor(val)
  avg_tensor = hvd.allreduce(tensor, name=name)
  return avg_tensor.item()
```

where `validation_steps` is derived by:

`validation_steps = val_df_size // BATCH_SIZE #  (the length of the data set // batch size)`

When training, I get results like this:

```
Validation Loss: 0.5668 Acc: 0.7325 F1: 0.0071
```

My BATCH_SIZE is 100, so they places the f1 where I suspect it, by undoing the multiplication, to around .71. 

But why is it that this works(?), if batch size is the culprit for my f1 metric.

```
epoch_acc = running_corrects.double() / (validation_steps * BATCH_SIZE)
```

**Comment by tgaddair** [comment]
Hey @afogarty85, can you clarify what it is you're trying to do?  One common way to collect metrics across workers is to use `hvd.allgather_object`, which collects ordinary Python objects as a list (one element per worker).  You can then do more complex aggregations from there.  It may be the case that averaging the f1 score is not the correct calculation.

**Comment by afogarty85** [comment]
I am trying to get an f1 score for my evaluation run, much like the accuracy score. Do you happen to have an example of this `hvd.allgather_object` running out in the wild? I see your test functions when searching internally.

**Comment by tgaddair** [comment]
Sure, we have an example in the Ludwig library here:

https://github.com/uber/ludwig/blob/master/ludwig/models/predictor.py#L241
